crates/async-sawtooth-sdk/src/ledger.rs

```diff
@@ -24,19 +31,53 @@ use sawtooth_sdk::messages::{
 };
 use serde::{Deserialize, Serialize};
 
-use tracing::{debug, error, instrument};
+use thiserror::Error;
+use tracing::{debug, error, instrument, trace};
 
 use crate::{
     error::SawtoothCommunicationError, sawtooth::MessageBuilder,
     zmq_client::RequestResponseSawtoothChannel,
 };
 
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
-pub struct BlockId(String);
+#[derive(Debug, Error)]
+pub enum BlockIdError {
+    #[error("Not hex")]
+    Hex(#[from] hex::FromHexError),
+
+    #[error("Not 32 bytes")]
+    Size(#[from] std::array::TryFromSliceError),
+}
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub enum BlockId {
+    Marker, //Block ids can be null, empty string etc
+    Block([u8; 32]),
+}
+
+impl TryFrom<String> for BlockId {
+    type Error = Infallible;
+
+    /// Sawtooth uses a special marker value for the first block, which is not
+    /// parsable as a hash - 8 hex zero bytes.
+    #[instrument(level = "trace", ret(Debug))]
+    fn try_from(s: String) -> Result<BlockId, Infallible> {
+        Ok(hex::decode(s)
+            .map_err(BlockIdError::from)
+            .and_then(|x| {
+                x.as_slice()
+                    .try_into()
+                    .map_err(BlockIdError::from)
+                    .map(BlockId::Block)
+            })
+            .unwrap_or(BlockId::Marker))
```

Comment:
Should we check they really are zeroes?

Reply:
The binary protocol using hex encoded strings for hashes in the first place leads down this dark path :)

Reply:
They are case sensitive too! I found that out the hard way